src/feature/relay/router.h

```diff
@@ -73,74 +65,129 @@ uint16_t router_get_advertised_dir_port(const or_options_t *options,
-                                                      const routerinfo_t *ri);
-STATIC void router_update_routerinfo_from_extrainfo(routerinfo_t *ri,
-                                                    const extrainfo_t *ei);
-STATIC int router_dump_and_sign_routerinfo_descriptor_body(routerinfo_t *ri);
 #endif /* defined(TOR_UNIT_TESTS) */
 
 #endif /* defined(ROUTER_PRIVATE) */
 
+#else /* !defined(HAVE_RELAY_MODE) */
+
+#define IGNORE1(a)                              \
+  ((void)(a))
+#define IGNORE1_RET(a, r)                       \
+  (IGNORE1(a), (r))
+#define IGNORE2(a,b)                            \
+  ((void)(a), (void)(b))
+#define IGNORE2_RET(a,b, r)                     \
+  (IGNORE2((a),(b)), (r))
+#define IGNORE3(a,b,c)                          \
+  ((void)(a), (void)(b), (void)(c))
+#define IGNORE3_RET(a,b,c, r)                   \
+  (IGNORE3((a),(b),(c)), (r))
+
+#define get_server_identity_key()(tor_abort_(),NULL)
+#define dup_onion_keys(key,last) \
+  (*(key)=NULL, *(last)=NULL)
+#define expire_old_onion_keys() \
+  STMT_NIL
+#define rotate_onion_key() \
+  STMT_NIL
+
+#define construct_ntor_key_map() \
+  tor_malloc(1)
+#define ntor_key_map_free(map) \
+  tor_free(map)
+
+#define router_digest_is_me(digest) \
+  IGNORE1_RET((digest), 0)
+#define router_get_my_id_digest()               \
+  ("\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"   \
+   "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0")
+#define router_extrainfo_digest_is_me(digest) \
+  IGNORE1_RET((digest), 0)
+#define router_is_me(r)                         \
+  IGNORE1_RET((r), 0)
+#define server_identity_key_is_set() \
+  (0)
+
+#define router_get_advertised_or_port(opts) \
+  IGNORE1_RET((opts), 0)
+#define router_get_advertised_or_port_by_af(opts, af) \
+  IGNORE2_RET((opts), (af), 0)
+#define router_get_advertised_dir_port(opts, port) \
+  IGNORE2_RET((opts), (port), 0)
+#define router_get_active_listener_port_by_type_af(lt,fam) \
+  IGNORE2_RET((lt), (fam), 0)
+
+#define router_compare_to_my_exit_policy(addr, port) \
+  IGNORE2_RET((addr), (port), -1)
+#define router_my_exit_policy_is_reject_star() \
+  (1)
+
+#define should_refuse_unknown_exits(opts) \
+  IGNORE1_RET((opts), 1)
+
+/* TODO: These should be disabled by v3auth mode, not relay mode. */
+#define get_my_v3_authority_cert()              \
+  NULL
+#define get_my_v3_authority_signing_key()       \
+  NULL
+#define get_my_v3_legacy_cert()                 \
+  NULL
+#define get_my_v3_legacy_signing_key()          \
+  NULL
+
+#define routerinfo_err_is_transient(e)          \
+  IGNORE1_RET((e), 0)
+#define routerinfo_err_to_string(e)             \
+  IGNORE1_RET((e), "Tor was compiled without relay support")
+
+#define router_pick_published_address(opt, addr, cache) \
+  IGNORE3_RET((opt), (addr), (cache), -1)
+
+#define router_new_address_suggestion(s, d) \
+  IGNORE2((s),(d))
+
+#define init_keys() \
+  init_keys_client()
+
+#define init_keys_common() \
+  (0)
```

Comment:
init_keys_common() creates the key mutex.

If we don't need the key mutex on clients, let's open another ticket to remove it?
(And if we do need it, let's leave it alone.)

Reply:
I think that key_lock *is* currently removed on clients with this patch.  It is declared in router.c, and is only used there: it protects onion keys only.

Reply:
Can we have some comments around these changes, for the sake of future developers? :-'